client/cgo/cgotest.go

```diff
@@ -0,0 +1,222 @@
+}
+
+func testVerifySignature(t *testing.T) {
+	key, err := sign.GenerateKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	message := []byte("test message")
+	sig := key.Sign(message)
+
+	pk, ok := key.Public()
+	if !ok {
+		t.Errorf("bad PK?")
+	}
+
+	if v := C.testVerifySignature(byteSliceToCpchar(pk), C.int(len(pk)),
+		byteSliceToCpchar(message), C.int(len(message)),
+		byteSliceToCpchar(sig), C.int(len(sig))); v != 1 {
+		t.Error("cgoVerifySignature failed")
+	}
+}
+
+func testHashChainInvalidSize(t *testing.T) {
+	if v := C.testHashChainInvalidSize(); v != 0 {
+		t.Error("Expect hash chain verification returns false")
+	}
+}
+
+func testVerifyAuthPath(t *testing.T) {
+	signKey, err := sign.GenerateKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, vrfPrivKey, err := vrf.GenerateKey(bytes.NewReader(
+		[]byte("deterministic tests need 256 bit")))
+	if err != nil {
+		t.Fatal(err)
+	}
+	pad, err := merkletree.NewPAD(merkletree.NewPolicies(2, vrfPrivKey), signKey, 10)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	key1 := "key1"
+	val1 := []byte("value1")
+	key2 := "key2"
+	val2 := []byte("value2")
+	key3 := "key3"
+	val3 := []byte("value3")
+
+	if err := pad.Set(key1, val1); err != nil {
+		t.Fatal(err)
+	}
+	if err := pad.Set(key2, val2); err != nil {
+		t.Fatal(err)
+	}
+	if err := pad.Set(key3, val3); err != nil {
+		t.Fatal(err)
+	}
+	pad.Update(nil)
+
+	// proof of inclusion
+	proof, err := pad.Lookup(key3)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	isLeafEmpty := 0
+	if proof.Leaf.IsEmpty() {
+		isLeafEmpty = 1
+	}
+	if v := C.testVerifyAuthPath(byteSliceToCpchar(pad.GetLatestSTR().Root()), C.int(len(pad.GetLatestSTR().Root())),
+		byteSliceToCpchar(proof.TreeNonce), C.int(len(proof.TreeNonce)),
+		byteSliceToCpchar(proof.LookupIndex), C.int(len(proof.LookupIndex)),
+		(**C.uchar)(unsafe.Pointer(&proof.PrunedTree[0][0])), C.int(len(proof.PrunedTree)), C.int(len(proof.PrunedTree[0])),
+		C.int(proof.Leaf.Level()),
+		byteSliceToCpchar(proof.Leaf.Index()), C.int(len(proof.Leaf.Index())),
+		byteSliceToCpchar(proof.Leaf.Commitment()), C.int(len(proof.Leaf.Commitment())),
+		C.int(isLeafEmpty)); v != 1 {
+		t.Error("Verify proof of inclusion failed")
+	}
+
+	// proof of absence
+	proof, err = pad.Lookup("123")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	isLeafEmpty = 0
+	if proof.Leaf.IsEmpty() {
+		isLeafEmpty = 1
+	}
+	if v := C.testVerifyAuthPath(byteSliceToCpchar(pad.GetLatestSTR().Root()), C.int(len(pad.GetLatestSTR().Root())),
+		byteSliceToCpchar(proof.TreeNonce), C.int(len(proof.TreeNonce)),
+		byteSliceToCpchar(proof.LookupIndex), C.int(len(proof.LookupIndex)),
+		(**C.uchar)(unsafe.Pointer(&proof.PrunedTree[0][0])), C.int(len(proof.PrunedTree)), C.int(len(proof.PrunedTree[0])),
+		C.int(proof.Leaf.Level()),
+		byteSliceToCpchar(proof.Leaf.Index()), C.int(len(proof.Leaf.Index())),
+		nil, C.int(0),
+		C.int(isLeafEmpty)); v != 1 {
+		t.Error("Verify proof of absence failed")
+	}
+}
+
+func testVerifyProofOfAbsenceSamePrefix(t *testing.T) {
```

Comment:
No, since this is done in ``merkletree.VerifyAuthPath()``, right? I mean, should we assert that ``proof.Leaf.IsEmpty()`` is false for "a" and true for "123" in ``testVerifyAuthPath()`` above.